lib/logflare/backends/adaptor/postgres_adaptor.ex

```diff
@@ -77,6 +69,42 @@ defmodule Logflare.Backends.Adaptor.PostgresAdaptor do
     |> Ecto.Changeset.validate_format(:url, ~r/postgresql?\:\/\/.+/)
   end
 
+  @impl true
+  def queryable?(), do: true
+
+  @impl true
+  def execute_query(%SourceBackend{} = source_backend, %Ecto.Query{} = query) do
+    mod = create_repo(source_backend)
+    :ok = connect_to_repo(source_backend)
+    result = mod.all(query)
+    {:ok, result}
+  end
+
+  def execute_query(%SourceBackend{} = source_backend, query_string)
+      when is_binary(query_string) do
+    mod = create_repo(source_backend)
+    :ok = connect_to_repo(source_backend)
+    result = Ecto.Adapters.SQL.query!(mod, query_string)
+
+    rows =
+      for row <- result.rows do
+        result.columns |> Enum.zip(row) |> Map.new()
+      end
+
+    {:ok, rows}
+  end
+
+  # expose PgRepo functions
+  defdelegate create_repo(source_backend), to: PgRepo
+  defdelegate connect_to_repo(source_backend), to: PgRepo
+  defdelegate table_name(source_or_source_backend), to: PgRepo
+  defdelegate create_log_events_table(source_backend), to: PgRepo
+  defdelegate create_log_events_table(source_backend, override_migrations), to: PgRepo
+  defdelegate rollback_migrations(source_backend), to: PgRepo
+  defdelegate drop_migrations_table(source_backend), to: PgRepo
+  defdelegate migrations_table_name(source_backend), to: PgRepo
+  defdelegate insert_log_event(source_backend, log_event), to: PgRepo
```

Comment:
I don't fully agree with this since it's a layer of indirection but I can live with it 😄

Reply:
Reason for this is because ideally each Adaptor should expose and implement only the same callbacks as per the `Adaptor` behaviour, so that all logic for setup/teardown/migrations are encapsulated within the `*Adaptor` module namespace, so that internal implementation can always change without external api change. Whatever is exposed at Adaptor-level is essentially the public api for the backend.

In the long run, it would make our maintenance and testing much easier as the number of adaptors grow, since expected behaviour for each backend is always the same.

I'm leaning towards using `setup_ingest/1`, `setup_querying/1`. `teardown/1`, `insert_log_event/2` callbacks to `Adaptor`, but would like to see how the BigQueryAdaptor turns out first before extending Adaptor more with the commonalities.

Reply:
Ah got it!!! Makes sense 👍👍